File: agentlens/sdk-node/src/redaction.ts
// Redacción de PII en cliente (paridad con sdk-python/redaction.py, E1-T04).
// Primera barrera de privacidad: nada de PII sale del proceso sin mascarar.

import type { Attributes, AttributeValue } from "@opentelemetry/api";

interface Pattern {
  name: string;
  regex: RegExp;
  token: string;
}

// Orden: patrones más específicos primero (IBAN/tarjeta antes que teléfono).
const DEFAULT_PATTERNS: Array<[string, RegExp, string]> = [
  ["email", /[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}/g, "[REDACTED_EMAIL]"],
  ["iban", /\b[A-Z]{2}\d{2}[A-Z0-9]{11,30}\b/g, "[REDACTED_IBAN]"],
  ["card", /\b(?:\d[ \-]?){13,19}\b/g, "[REDACTED_CARD]"],
  [
    // Conservador para evitar falsos positivos sobre números sueltos: exige
    // prefijo internacional, área entre paréntesis o tres grupos separados.
    "phone",
    /(?:\+\d{1,3}[ \-]?\d{2,4}(?:[ \-]?\d{2,4}){1,4}|\(\d{1,4}\)[ \-]?\d{3,4}[ \-]?\d{3,4}|\d{3}[ \-]\d{3}[ \-]\d{2,4})/g,
    "[REDACTED_PHONE]",
  ],
];

export class Redactor {
  private patterns: Pattern[];

  constructor(enabled?: string[]) {
    this.patterns = DEFAULT_PATTERNS.filter(
      ([name]) => !enabled || enabled.includes(name),
    ).map(([name, regex, token]) => ({ name, regex, token }));
  }

  /** Añade un patrón propio (p.ej. nº de empleado). */
  addPattern(name: string, regex: RegExp, token: string): void {
    this.patterns.push({ name, regex, token });
  }

  redactText(text: string): string {
    let out = text;
    for (const p of this.patterns) {
      out = out.replace(p.regex, p.token);
    }
    return out;
  }

  redactValue(value: AttributeValue): AttributeValue {
    if (typeof value === "string") {
      return this.redactText(value);
    }
    if (Array.isArray(value)) {
      return value.map((v) => (typeof v === "string" ? this.redactText(v) : v)) as AttributeValue;
    }
    return value;
  }

  redactAttributes(attributes: Attributes): Attributes {
    const out: Attributes = {};
    for (const [k, v] of Object.entries(attributes)) {
      out[k] = v === undefined ? v : this.redactValue(v);
    }
    return out;
  }
}

File: Agentlens/agentlens/sdk-python/src/agentlens/redaction.py
"""Redacción de PII en el lado del cliente.

Esta es la primera de las dos barreras de privacidad (la segunda es Presidio en
el Collector). El objetivo es que ningún dato personal salga del proceso del
agente sin mascarar, cumpliendo el principio de minimización (GDPR Art. 5).

El núcleo es puro y testeable: ``Redactor.redact_text`` no tiene efectos
secundarios y es trivial de cubrir con tests.
"""
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Iterable, List, Tuple

# Orden importa: los patrones más específicos van primero para evitar que un
# patrón genérico (teléfono) "se coma" parte de uno específico (IBAN, tarjeta).
_DEFAULT_PATTERNS: List[Tuple[str, str, str]] = [
    (
        "email",
        r"[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}",
        "[REDACTED_EMAIL]",
    ),
    (
        "iban",
        r"\b[A-Z]{2}\d{2}[A-Z0-9]{11,30}\b",
        "[REDACTED_IBAN]",
    ),
    (
        # 13-19 dígitos, opcionalmente separados por espacios o guiones.
        "card",
        r"\b(?:\d[ \-]?){13,19}\b",
        "[REDACTED_CARD]",
    ),
    (
        # Conservador: requiere prefijo + o paréntesis para reducir falsos
        # positivos sobre números normales.
        "phone",
        r"(?:\+\d{1,3}[ \-]?)?(?:\(\d{1,4}\)[ \-]?)?\d{3,4}[ \-]?\d{3,4}[ \-]?\d{0,4}",
        "[REDACTED_PHONE]",
    ),
]


@dataclass
class Pattern:
    name: str
    regex: "re.Pattern"
    token: str


class Redactor:
    """Aplica un conjunto de patrones de redacción sobre texto y atributos."""

    def __init__(self, patterns: Iterable[Tuple[str, str, str]] = _DEFAULT_PATTERNS,
                 enabled: Iterable[str] | None = None):
        self._patterns: List[Pattern] = []
        enabled_set = set(enabled) if enabled is not None else None
        for name, rgx, token in patterns:
            if enabled_set is not None and name not in enabled_set:
                continue
            self._patterns.append(Pattern(name, re.compile(rgx), token))

    def add_pattern(self, name: str, regex: str, token: str) -> None:
        """Permite a un cliente añadir patrones propios (p.ej. nº de empleado)."""
        self._patterns.append(Pattern(name, re.compile(regex), token))

    def redact_text(self, text: str) -> str:
        for p in self._patterns:
            text = p.regex.sub(p.token, text)
        return text

    def redact_value(self, value):
        """Redacta solo valores de tipo string; otros tipos pasan intactos."""
        if isinstance(value, str):
            return self.redact_text(value)
        if isinstance(value, (list, tuple)):
            return type(value)(self.redact_value(v) for v in value)
        return value

    def redact_attributes(self, attributes: dict) -> dict:
        """Devuelve un nuevo dict con los valores redactados (no muta el original)."""
        return {k: self.redact_value(v) for k, v in attributes.items()}
